gateway/app/drivers/quality_on_demand/nef/nef_interface.py

```diff
@@ -0,0 +1,429 @@
+            str(sink),
+            content=cloud_event.model_dump_json(exclude_unset=True, by_alias=True),
+        )
+
+        await self._save_subscription_info(id, sub_info)
+
+    def _get_qod_status(self, nef_response: UserPlaneNotificationData) -> Status | None:
+        status = None
+        for report in nef_response.eventReports:
+            if report.event == UserPlaneEvent.QOS_GUARANTEED:
+                status = Status.AVAILABLE
+            elif report.event == UserPlaneEvent.FAILED_RESOURCES_ALLOCATION:
+                status = Status.UNAVAILABLE
+            elif report.event == UserPlaneEvent.QOS_NOT_GUARANTEED:
+                status = Status.REQUESTED
+            elif report.event == UserPlaneEvent.SUCCESSFUL_RESOURCES_ALLOCATION:
+                status = Status.AVAILABLE
+        return status
+
+    async def _save_subscription_info(self, id: str, data: SessionInfo) -> None:
+        key = f"{_prefix_info}:{id}"
+
+        await self.redis.set(
+            key, data.model_dump_json(exclude_unset=True, by_alias=True)
+        )
+
+    def _get_flow_description(self, req: CreateSession | SessionInfo) -> List[FlowInfo]:
+        flow = []
+
+        device_ports = ""
+        device_ports_range = ""
+        if req.devicePorts:
+            if req.devicePorts.ports:
+                device_ports = str(req.devicePorts.ports[0])
+                for port in req.devicePorts.ports[1:]:
+                    device_ports += f",{port}"
+            if req.devicePorts.ranges:
+                if device_ports:
+                    device_ports_range += ","
+                device_ports_range += (
+                    f"{req.devicePorts.ranges[0].from_}-{req.devicePorts.ranges[0].to}"
+                )
+                for ranges in req.devicePorts.ranges[1:]:
+                    device_ports_range += f",{ranges.from_}-{ranges.to}"
+
+        server_ports = ""
+        server_ports_range = ""
+
+        if req.applicationServerPorts:
+            if req.applicationServerPorts.ports:
+                server_ports = str(req.applicationServerPorts.ports[0])
+                for port in req.applicationServerPorts.ports[1:]:
+                    server_ports += f",{port}"
+            if req.applicationServerPorts.ranges:
+                if server_ports:
+                    server_ports_range += ","
+                server_ports_range += f"{req.applicationServerPorts.ranges[0].from_}-{req.applicationServerPorts.ranges[0].to}"
+                for ranges in req.applicationServerPorts.ranges[1:]:
+                    server_ports_range += f",{ranges.from_}-{ranges.to}"
+
+        dev_ips = []
+        app_ips = []
+
+        if req.device and req.device.ipv4Address is not None:
+            dev_ips.append(str(req.device.ipv4Address.publicAddress))
+            app_ips.append(req.applicationServer.ipv4Address)
+
+        if req.device and req.device.ipv6Address:
+            dev_ips.append(req.device.ipv6Address.exploded)
```

Comment:
```suggestion
            dev_ips.append(req.device.ipv6Address)
```

You don't need exploded here.